sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/PushNotificationClient.java

```diff
@@ -322,30 +324,44 @@ private void refreshEncryptionKeys() throws Throwable {
 
         this.cryptoKey = this.keyGenerator.generateKey();
         this.authKey = this.keyGenerator.generateKey();
+
+        this.keyRotateTimeMillis = System.currentTimeMillis();
     }
 
     private String decryptPayload(String encryptedPayload) throws Throwable {
-        this.logger.info(" Decrypting payload.");
+        this.logger.verbose(" Decrypting payload.");
 
         byte[] encryptedBytes = Base64Util.decodeString(encryptedPayload);
-        byte[] f = NotificationUtils.extractEncryptionKey(encryptedBytes);
+        byte[] encryptionKey = NotificationUtils.extractEncryptionKey(encryptedBytes);
         byte[] iv = NotificationUtils.extractInitializationVector(encryptedBytes);
         byte[] cipherText = NotificationUtils.extractCipherText(encryptedBytes);
         byte[] hmac = NotificationUtils.extractHmac(encryptedBytes);
 
-        if (NotificationUtils.verifyEncryptedPayload(f, iv, cipherText, hmac, this.authKey)) {
+        if (NotificationUtils.verifyEncryptedPayload(encryptionKey, iv, cipherText, hmac, this.authKey)) {
             return NotificationUtils.decryptPushNotificationPayload(iv, cipherText, this.cryptoKey);
         }
 
         // When client has registered a new key, because of eventual consistency, latencies and concurrency,
         // the old key can still be used by server side for some notifications
         // Try old key if the new key failed to decrypt the payload.
-        if (NotificationUtils.verifyEncryptedPayload(f, iv, cipherText, hmac, this.previousAuthKey)) {
+        if (inKeyRotationGracePeriod()
+            && NotificationUtils.verifyEncryptedPayload(encryptionKey, iv, cipherText, hmac, this.previousAuthKey)) {
             return NotificationUtils.decryptPushNotificationPayload(iv, cipherText, this.previousCryptoKey);
         }
 
         // Reject the push when computed signature does not match the included signature - it can not be trusted
         throw logger.logExceptionAsError(
             new RuntimeException("Invalid encrypted push notification payload! Drop the request."));
```

Comment:
I think we should throw an exception with a different error message if the problem was due to an invalid key, as we really don't know if the payload was valid or not and saying so is misleading for users and makes things hard to debug.